src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,463 @@
+
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        """
+        Evaluates the passed callable directly using values of Ad_array instances
+        passed as arguments.
+
+        :param args: tuple of :class:`~porepy.numerics.ad.forward_mode.Ad_array`
+        :type args: tuple
+        :return: results from callable packed into an array.
+        :rtype: numpy.array
+        """
+        # get values of argument Ad_arrays.
+        vals = (arg.val for arg in args)
+
+        # if the callable is flagged as conform for vector operations, feed vectors
+        if self.is_vector_func:
+
+            return self.func(*vals)
+        else:
+            # if not vector-conform, feed element-wise
+
+            # TODO this displays some special behavior when val-arrays have different lengths:
+            # it returns None-like things for every iteration more then shortest length
+            # These Nones are ignored for some reason by the function call, as well as by the
+            # array constructor.
+            # If a mortar var and a subdomain var are given as args,
+            # then the lengths will be different for example.
+            return np.array([self.func(*vals_i) for vals_i in zip(*vals)])
+
+
+### CONCRETE IMPLEMENTATIONS ------------------------------------------------------------------
+
+
+class Function(AbstractFunction):
+    """Ad representation of an analytically given function,
+    where it is expected that feeding the callable with Ad_arrays directly will give
+    give the proper result.
+
+    Here the values AND the Jacobian are obtained exactly by standard rules of differentiation.
+
+    The intended use is as a wrapper for operations on pp.ad.Ad_array objects,
+    in forms which are not directly or easily expressed by the rest of the Ad
+    framework.
+
+    NOTE: this is a special case where `get_values` and `get_jacobian` are formally implemented
+    but never used by the AD framework
+    (due to the assumed, direct evaluation of result using the callable).
+    """
+
+    def __init__(
+        self, func: Callable, name: str, is_vector_func: Optional[bool] = True
+    ):
+        super().__init__(func, name, is_vector_func, True)
+
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.val
+
+    def get_jacobian(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.jac
+
+
+class LJacFunction(AbstractJacFunction):
+    """
+    Approximates the Jacobian of the function using the L-scheme
+    with a fixed value per dependency.
+    """
+
+    def __init__(
+        self,
+        L: Union[List[float], float],
+        func: Callable,
+        name: str,
+        is_vector_func: Optional[bool] = False,
+    ):
+        """Constructor.
+
+        The L-multiplier for the L-scheme can be passed for every argument specifically
+        using a list.
```

Comment:
Again, the L-method may be a motivation, but there is no reason to make a class dedicated to this.